delivery_carrier_label_dispatch/wizard/generate_labels.py

```diff
@@ -40,51 +50,207 @@ def _get_dispatch_ids(self, cr, uid, context=None):
     _columns = {
         'dispatch_ids': fields.many2many('picking.dispatch',
                                          string='Picking Dispatch'),
-        'label_pdf_file': fields.binary('Labels file'),
+        'generate_new_labels': fields.boolean(
+            'Generate new labels',
+            help="If this option is used, new labels will be     "
+                 "generated for the packs even if they already have one.\n"
+                 "The default is to use the existing label."),
     }
 
     _defaults = {
         'dispatch_ids': _get_dispatch_ids,
+        'generate_new_labels': False,
     }
 
+    def _get_packs(self, cr, uid, wizard, dispatch, context=None):
+        moves = sorted(dispatch.move_ids, key=attrgetter('tracking_id.name'))
+        for pack, moves in groupby(moves, key=attrgetter('tracking_id')):
+            pack_label = self._find_pack_label(cr, uid, wizard, pack,
+                                               context=context)
+            yield pack, list(moves), pack_label
+
+    def _find_picking_label(self, cr, uid, wizard, picking, context=None):
+        label_obj = self.pool['shipping.label']
+        domain = [('file_type', '=', 'pdf'),
+                  ('res_id', '=', picking.id),
+                  ('tracking_id', '=', False),
+                  ]
+        label_id = label_obj.search(cr, uid, domain, order='create_date DESC',
+                                    limit=1, context=context)
+        if not label_id:
+            return None
+        return label_obj.browse(cr, uid, label_id[0], context=context)
+
+    def _find_pack_label(self, cr, uid, wizard, pack, context=None):
+        label_obj = self.pool['shipping.label']
+        domain = [('file_type', '=', 'pdf'),
+                  ('tracking_id', '=', pack.id),
+                  ]
+        label_id = label_obj.search(cr, uid, domain, order='create_date DESC',
+                                    limit=1, context=context)
+        if not label_id:
+            return None
+        return label_obj.browse(cr, uid, label_id[0], context=context)
+
+    def _do_generate_labels(self, cr, uid, wizard, pack, picking,
+                            label, context=None):
+        """ Generate a label in a thread safe context
+
+        Here we declare a specific cursor so do not launch
+        too many threads
+        """
+        picking_out_obj = self.pool['stock.picking.out']
+        # generate the label of the pack
+        tracking_ids = [pack.id] if pack else None
+        # create a cursor to be thread safe
+        thread_cr = pooler.get_db(cr.dbname).cursor()
```

Comment:
Not sure how to catch this.

By the way wouldn't this kind of error block generation if all workers where killed?